File: LeetCode/Medium/Combinations/Go/combinations.go
package main

import "fmt"

func combine(n int, k int) [][]int {
	res := [][]int{}
	backtrack(n, k, 1, []int{}, &res)
	return res
}

func backtrack(n, k, start int, temp []int, res *[][]int) {
	if k == 0 {
		*res = append(*res, append([]int{}, temp...))
		return
	}
	for i := start; i <= n; i++ {
		temp = append(temp, i)
		backtrack(n, k-1, i+1, temp, res)
		temp = temp[:len(temp)-1]
	}
}

func main() {
	fmt.Println(combination(4, 2))
}

File: LeetCode/Medium/Combinations/Java/combinations.java
class Solution {
    List<List<Integer>> res;
    public List<List<Integer>> combine(int n, int k) {
        res = new ArrayList<>();
        backtrack(n, k, new ArrayList<>(), 1);
        return res;
    }

    private void backtrack(int n, int k, List<Integer> list, int start) {
        if (k == 0) {
            res.add(new ArrayList<>(list));
            return;
        }
        for (int i=start; i<=n; i++) {
            list.add(i);
            backtrack(n, k-1, list, i+1);
            list.remove(list.size()-1);
        }
    }
}
